plugin/evm/vm_warp_test.go

```diff
@@ -120,6 +130,262 @@ func TestSendWarpMessage(t *testing.T) {
+		big.NewInt(0),
+	)
+	require.NoError(err)
+
+	testWarpVMTransaction(t, unsignedMessage, false, exampleWarpPayload)
+}
+
+func testWarpVMTransaction(t *testing.T, unsignedMessage *avalancheWarp.UnsignedMessage, validSignature bool, txPayload []byte) {
+	require := require.New(t)
+	genesis := &core.Genesis{}
+	require.NoError(genesis.UnmarshalJSON([]byte(genesisJSONDUpgrade)))
+	genesis.Config.GenesisPrecompiles = params.Precompiles{
+		warp.ConfigKey: warp.NewDefaultConfig(subnetEVMUtils.NewUint64(0)),
+	}
+	genesisJSON, err := genesis.MarshalJSON()
+	require.NoError(err)
+	issuer, vm, _, _ := GenesisVM(t, true, string(genesisJSON), "", "")
+
+	defer func() {
+		require.NoError(vm.Shutdown(context.Background()))
+	}()
+
+	acceptedLogsChan := make(chan []*types.Log, 10)
+	logsSub := vm.eth.APIBackend.SubscribeAcceptedLogsEvent(acceptedLogsChan)
+	defer logsSub.Unsubscribe()
+
+	nodeID1 := ids.GenerateTestNodeID()
+	blsSecretKey1, err := bls.NewSecretKey()
+	require.NoError(err)
+	blsPublicKey1 := bls.PublicFromSecretKey(blsSecretKey1)
+	blsSignature1 := bls.Sign(blsSecretKey1, unsignedMessage.Bytes())
+
+	nodeID2 := ids.GenerateTestNodeID()
+	blsSecretKey2, err := bls.NewSecretKey()
+	require.NoError(err)
+	blsPublicKey2 := bls.PublicFromSecretKey(blsSecretKey2)
+	blsSignature2 := bls.Sign(blsSecretKey2, unsignedMessage.Bytes())
+
+	blsAggregatedSignature, err := bls.AggregateSignatures([]*bls.Signature{blsSignature1, blsSignature2})
+	require.NoError(err)
+
+	minimumValidPChainHeight := uint64(10)
+	getValidatorSetTestErr := errors.New("can't get validator set test error")
+
+	vm.ctx.ValidatorState = &validators.TestState{
+		// TODO: test both Primary Network / C-Chain and non-Primary Network
+		GetSubnetIDF: func(ctx context.Context, chainID ids.ID) (ids.ID, error) {
+			return ids.Empty, nil
+		},
+		GetValidatorSetF: func(ctx context.Context, height uint64, subnetID ids.ID) (map[ids.NodeID]*validators.GetValidatorOutput, error) {
+			if height < minimumValidPChainHeight {
+				return nil, getValidatorSetTestErr
+			}
+			return map[ids.NodeID]*validators.GetValidatorOutput{
+				nodeID1: {
+					NodeID:    nodeID1,
+					PublicKey: blsPublicKey1,
+					Weight:    50,
+				},
+				nodeID2: {
+					NodeID:    nodeID2,
+					PublicKey: blsPublicKey2,
+					Weight:    50,
+				},
+			}, nil
+		},
+	}
+
+	signersBitSet := set.NewBits()
+	signersBitSet.Add(0)
+	signersBitSet.Add(1)
+
+	warpSignature := &avalancheWarp.BitSetSignature{
+		Signers: signersBitSet.Bytes(),
+	}
+
+	blsAggregatedSignatureBytes := bls.SignatureToBytes(blsAggregatedSignature)
+	copy(warpSignature.Signature[:], blsAggregatedSignatureBytes)
+
+	signedMessage, err := avalancheWarp.NewMessage(
+		unsignedMessage,
+		warpSignature,
+	)
+	require.NoError(err)
+
+	createTx, err := types.SignTx(
+		types.NewContractCreation(0, common.Big0, 7_000_000, big.NewInt(225*params.GWei), common.Hex2Bytes(exampleWarpBin)),
+		types.LatestSignerForChainID(vm.chainConfig.ChainID),
```

Comment:
can we have these as constants?

Reply:
Three is the magic number, so going to leave this as is for now.